yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryProvider.scala

```diff
@@ -0,0 +1,1015 @@
+    this.synchronized {
+      applications = newVal
+    }
+  }
+
+  /**
+   * Health check to call before any other operation is attempted.
+   * This is atomic, using the `healthy` flag to check.
+   * If the endpoint is considered unhealthy then the healthy flag
+   * is reset to false and an exception thrown.
+   * @return true if the health check took place
+   */
+  protected def maybeCheckHealth(): Boolean = {
+    val h = getHealthFlag();
+    if (!h.getAndSet(true)) {
+      val client = getTimelineQueryClient()
+      try {
+        client.healthCheck()
+        true
+      } catch {
+        case e: Exception =>
+          // failure
+          logWarning(s"Health check of $client failed", e)
+          setLastFailure(e)
+          // reset health so another caller may attempt it.
+          h.set(false)
+          // propagate the failure
+          throw e;
+      }
+    } else {
+      false
+    }
+  }
+
+  /**
+   * Start the refresh thread with the given interval.
+   *
+   * When this thread exits, it will close the `timelineQueryClient`
+   * instance
+   */
+  def startRefreshThread(): Unit = {
+    logInfo(s"Starting timeline refresh thread")
+    val thread = new Thread(refresher, s"YarnHistoryProvider Refresher")
+    thread.setDaemon(true)
+    refresher.start(thread)
+  }
+
+  /**
+   * Stop the refresh thread if there is one.
+   *
+   * This does not guarantee an immediate halt to the thread.
+   * @return true if there was a refresh thread to stop
+   */
+  def stopRefreshThread(): Boolean = {
+    refresher.stopRefresher()
+  }
+
+  /**
+   * Probe for the refresh thread running
+   * @return true if the refresh thread has been created and is still alive
+   */
+  def isRefreshThreadRunning(): Boolean = {
+    refresher.isRunning()
+  }
+
+  def getRefreshCount(): Long = { refreshCount.get() }
+  def getRefreshFailedCount(): Long = { refreshFailedCount.get() }
+
+  /**
+   * List applications.
+   * <p>
+   * If the timeline is not enabled, returns an empty list
+   * @return  the result of the last successful listing operation,
+   *          or the `emptyListing` result if no listing has yet been successful
+   */
+   def listApplications(limit: Option[Long] = None,
+      windowStart: Option[Long] = None,
+      windowEnd: Option[Long] = None): ApplicationListingResults = {
+    if (!enabled) {
+      // Timeline is disabled: return the empty listing
+      return emptyListing
+    }
+    try {
+      maybeCheckHealth()
+      val client = getTimelineQueryClient()
+      logInfo(s"getListing from: $client")
+      // get the timestamp after any health check
+      val timestamp = now()
+      val timelineEntities =
+        client.listEntities(YarnHistoryService.SPARK_EVENT_ENTITY_TYPE,
+          windowStart = windowStart,
+          windowEnd = windowEnd,
+          limit = limit)
+
+      val listing = timelineEntities.flatMap { en =>
```

Comment:
Does the ATS guarantee ordering of the results? Does it match the order expected by the history server? (Or maybe you handle that somewhere else, there's still a lot of code to go through here...)